File: heap_sort/heap_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

void heapify(vector<int> &arr, int N, int i) {

    int largest = i;

    int l = 2 * i + 1;

    int r = 2 * i + 2;

    if (l < N && arr[l] > arr[largest])
        largest = l;

    if (r < N && arr[r] > arr[largest])
        largest = r;

    if (largest != i) {
        swap(arr[i], arr[largest]);

        heapify(arr, N, largest);
    }
}

vector<int> heapSort(vector<int> &arr, int N) {

    for (int i = N / 2 - 1; i >= 0; i--)
        heapify(arr, N, i);

    for (int i = N - 1; i > 0; i--) {

        swap(arr[0], arr[i]);

        heapify(arr, i, 0);
    }
    return arr;
}

File: heap_sort/heap_sort.go
package heap_sort

import "fmt"

func HeapSort() {
	arr := []int{2, 2, 2, 2, 2, 2, 2, 2, 2, 2, -9, 5, 11, 13, 12}
	heapSort(arr, len(arr))

	fmt.Println("Heap  Sort:", arr)
}

func heapSort(arr []int, n int) {
	for i := n/2 - 1; i >= 0; i-- {
		heapify(arr, n, i)
	}

	for i := n - 1; i > 0; i-- {
		arr[0], arr[i] = arr[i], arr[0]
		heapify(arr, i, 0)
	}
}

func heapify(arr []int, n int, i int) {
	largest := i
	l := 2*i + 1
	r := 2*i + 2

	if l < n && arr[l] > arr[largest] {
		largest = l
	}

	if r < n && arr[r] > arr[largest] {
		largest = r
	}

	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)
	}
}
